docs/rfc/rfc-002-zero-copy-encoding.md

```diff
@@ -0,0 +1,402 @@
+
+### New Protobuf Linting and Breaking Change Rules
+
+This zero-copy encoding places some additional requirements on the definition and maintenance of protobuf schemas.
+
+#### No New Fields Can Be Added To Existing Messages
+
+The biggest change is that it will be invalid to add a new field to an existing message and a breaking change detector
+will need to be created which augments [buf breaking](https://docs.buf.build/breaking/overview) to detect this.
+
+The reasons for this are two-fold:
+
+1) from an API compatibility perspective, adding a new field to an existing message is actually a state machine breaking
+   change which in [ADR 020](../architecture/adr-020-protobuf-transaction-encoding.md) required us to add an unknown
+   field detector. Furthermore, in [ADR 054](../architecture/adr-054-semver-compatible-modules.md) this "feature" of protobuf
+   poses one of the biggest problems for correct forward compatibility between different versions of the same module.
+2) not allowing new fields in existing messages makes the generated code in languages like Rust (which is currently our
+   highest priority target), much simpler and more performant because we can assume a fixed size struct gets allocated.
+   If new fields can be added to existing messages, we need to encode the number of fields into the message and then
+   do runtime checks. So this both increases memory layers and requires another layout of indirection. With the encoding
+   proposed below, "plain old Rust structs" (used with some special field types) can be used.
+
+Instead of adding new fields to existing messages, APIs can add new messages to existing packages or create new packages
+with new versions of the messages. Also, we are not restricting the addition of cases to `oneof`s or values to `enum`s.
+All of these cases are easier to detect at runtime with standard `switch` statements than the addition of new fields.
+
+#### Additional Linting Rules
+
+The following additional rules will be enforced by a linter that
+complements [buf lint](https://docs.buf.build/lint/overview):
+
+* all message fields must be specified in continuous ascending order starting from `1`
+* all enums must be specified in continuous ascending order starting from `0` - otherwise it is too complex to check at
+  runtime whether an enum value is unknown. An alternative would be to make adding new values to existing enums breaking
+* all enum values must be `<= 255`. Any enum in a blockchain application which needs more than 256 values is probably
+  doing something very wrong.
+* all oneof's must be the *only* element in their containing message and must start at field number `1` and be added in
+  continuous ascending order - this makes it possible to quickly check for unknown values
+* all `oneof` field numbers must be `<= 255`. Any `oneof` which needs more field cases is probably doing something very
+  wrong.
+
+These requirements make the encoding and generated code simpler.
```

Comment:
They also mean that you're no longer compatible with protocol buffers, and can't use the .proto extension, or the protobuf MIME type, and so on.

Reply:
That's not true

Reply:
OK, but you can't enforce that your code generator rules are followed by other consumers of your .proto files, so you can't assume any of the "additional requirements on the definition and maintenance of protobuf schemas" you define here will be followed. Is this not a problem?

Reply:
I don't think so. All proto files would originate as part of module APIs. We generally don't use proto files that originate from other sources. But even so, as I hinted at here: https://github.com/cosmos/cosmos-sdk/pull/15404#discussion_r1152551777, we can also generate code that doesn't follow these rules to be compatible, it would just be less optimized